### PR TITLE
Fixing up project compilation

### PR DESCRIPTION
**Pull request by adamhutchings**

### CODE REVIEW

**Comment by adamhutchings** [comment]
Closes #114, #112, #108, #103.

File: src/gen/gen.c [hunk 1]
```diff
@@ -232,6 +233,26 @@ static void yf_gen_if(
...
 int yfg_gen(struct yf_compile_analyse_job * data, struct yf_gen_info * info) {
 
+    /**
+     * Because we create the output file in the form of 'bin/c/foo/bar/baz.yf',
+     * we need to first mkdir() the enclosing directory, if it doesn't already
+     * exist. To do that, we would replace the last slash with a \0, mkdir() the
+     * path, and set it back.
+     * However, we have to create ALL enclosing folders first, so we need to get
+     * all slashes from left to right, and do this recursively.
+     */
```
Comment (dekrain) [inline src/gen/gen.c:243]: I think this could be moved to a separate function.

File: src/driver/compiler-backend.c [hunk 2]
```diff
@@ -73,34 +73,38 @@ static int create_output_file_name(
...
-            strcmp(data->output_file + strlen(data->output_file) - 3, ".yf")) {
+        int len = strlen(data->file_name);
+        /* Change .yf to .c */
+        /* FIRST, check for .yf ending */
+        if (len < 3 ||
+            strcmp(data->file_name + len - 3, ".yf") != 0) {
             /* Simply tack on .c */
             /* But no .yf is bad */
+            strcat(data->output_file, ".c");
             YF_PRINT_WARNING(
```
Comment (dekrain) [inline src/driver/compiler-backend.c:103]: I think the concatenation and message are in the wrong order.

File: src/gen/gen.c [hunk 3]
```diff
@@ -102,14 +104,27 @@ static void yf_gen_funcdecl(
...
+    if (strcmp(node->name->fn.name, "main")) {
+        fprintf(
+            out, "%s /* %s */ %s$$%s",
+            typebuf,
+            node->name->fn.rtype->name,
+            i->gen_prefix,
+            node->name->fn.name
+        );
+    } else {
+        fprintf(out, "int main");
```
Comment (dekrain) [inline src/gen/gen.c:124]: This doesn't check return type yet.

File: src/driver/compile.c [hunk 4]
```diff
@@ -121,6 +124,9 @@ int yf_run_compiler(struct yf_args * args) {
                 break;
 
             case YF_COMPILATION_EXEC:
+                /* Extra sanity check. */
+                if (!args->run_c_comp)
+                    break;
```
Comment (dekrain) [inline src/driver/compile.c:129]: I don't think this check is needed anymore?

File: src/driver/compile.c [hunk 5]
```diff
@@ -399,14 +418,9 @@ static int yfc_validate_compile(
...
-    if (yf_ensure_entry_point(pdata)) {
-        /* TODO - a more helpful error messsage */
-        YF_PRINT_ERROR("Expected exactly 1 'main' function.");
-        return 1;
-    }
-
-    if (adata->stage >= YF_COMPILE_CODEGEN)
+    if (adata->stage >= YF_COMPILE_CODEGENONLY) {
         retval = yf_backend_generate_code(adata);
+    }
```
Comment (dekrain) [inline src/driver/compile.c:423]: I think you want to add the check inside the if before generating code like
```
if (adata->stage == YF_COMPILE_FULL && yf_ensure_entry_point(pdata))
    return 1;
```

**Review by dekrain — CHANGES_REQUESTED** [review]

File: src/driver/compiler-backend.c (hunk 2 above)
Comment (adamhutchings) [reply src/driver/compiler-backend.c:103]: Very right.

File: src/gen/gen.c (hunk 3 above)
Comment (adamhutchings) [reply src/gen/gen.c:124]: Correct. It's to be done later.

File: src/gen/gen.c (hunk 1 above)
Comment (adamhutchings) [reply src/gen/gen.c:243]: Definitely agree.

File: src/driver/compile.c (hunk 5 above)
Comment (adamhutchings) [reply src/driver/compile.c:425]: Done.

**Review by dekrain — APPROVED** [review]

**Pull request merged**